use new containers/common/libimage package

vrothberg opened the pull request:
Move all code related to handling container images over to the new
`libimage` package in containers/common.  The new package is an
attempt to consolidate the code across the containers tools under the
github.com/containers umbrella.

The new `libimage` package provides functionality to perform all kinds
of operations for managing images such as local lookups, pushing,
pulling, listing, removing, etc.

The following packages have been moved over to containers/common:

`manifests` -> `common/libimage/manifests`
`pkg/manifests` -> `common/pkg/manifests`
`pkg/supplemented` -> `common/pkg/supplemented`

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

rhatdan commented:
@vrothberg needs rebase.

vrothberg commented:
> @vrothberg needs rebase.

Want to get it green first :)  Getting closer.

vrothberg commented:
Finally green. @rhatdan @nalind PTAL.

This is a _gigantic_ change but I did not find a way to split it into reasonable chunks.  Once https://github.com/containers/common/pull/497 is merged, I'll update the PR to vendor it the commit from c/common.

I moved as much code as I could find over to use `common/libimage`.  We may find more candidates in the future but I think we're good for now.

rhatdan commented on cmd/buildah/commit.go:
```diff
@@ -11,7 +11,8 @@ import (
 	"github.com/containers/common/pkg/auth"
-	"github.com/containers/image/v5/storage"
+	"github.com/containers/image/v5/pkg/shortnames"
+	storageTransport "github.com/containers/image/v5/storage"
```
Why are you doing this change?

vrothberg replied:
Some of our packages have the same name. In this case, `storage` could very well mean `containers/storage` and I find it sometimes hard to brain-parse code with all that ambiguity. That's why I used `storageTransport` or `dockerTransport` import aliases in `libimage` in hope to make code easier to read.

It's also the reason why I called it `libimage` and not `image`. This way, readers won't confuse it with `containers/image`.

rhatdan commented on cmd/buildah/images.go:
```diff
@@ -140,15 +121,28 @@ func imagesCmd(c *cobra.Command, args []string, iopts *imageResults) error {
+		return err
+	}
+
+	ctx := context.Background()
```
Nit: Why create a variable to use it once?

rhatdan commented:
@giuseppe @nalind @TomSweeneyRedHat @ashley-cui @umohnani8 PTAL

vrothberg commented:
Cc @baude @mheon

vrothberg commented:
Moving back to work-in-progress as I will start tackling Podman now which will likely require some changes to the libimage API.

vrothberg commented:
@rhatdan @TomSweeneyRedHat rebased and ready for your hungry eyes

nalind commented:
buildah is also a library, and this currently changes the signatures of exported functions, renames others, and removes exported types, functions, and packages entirely.  It will break consumers of the library.

vrothberg commented:
> buildah is also a library, and this currently changes the signatures of exported functions, renames others, and removes exported types, functions, and packages entirely. It will break consumers of the library.

Which consumers are you most concerned about?

nalind commented:
> > buildah is also a library, and this currently changes the signatures of exported functions, renames others, and removes exported types, functions, and packages entirely. It will break consumers of the library.
> 
> Which consumers are you most concerned about?

The first one that comes to mind is OpenShift's builder, but the concern is a general one that we have as a library.

vrothberg commented:
Yes that is a general issue but that’s an inevitable part of the code
consolidation. We need to do the same with libpod/Podman and eventually
CRI-O.

If changing parts of the external APIs is a deal breaker, I wished to have
known earlier; moving packages and the extend of the changes was discussed
a bunch of times.

@rhatdan, @nalind, I let you folks decide but I may need medication if I
have to thrash this PR.

A possible alternative solution is to cut a release branch now and then
merge the PR targeting a bump to v2.

On Thu 22 Apr 2021 at 18:20, Nalin Dahyabhai ***@***.***>
wrote:

> buildah is also a library, and this currently changes the signatures of
> exported functions, renames others, and removes exported types, functions,
> and packages entirely. It will break consumers of the library.
>
> Which consumers are you most concerned about?
>
> The first one that comes to mind is OpenShift's builder, but the concern
> is a general one that we have as a library.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/buildah/pull/3148#issuecomment-824987559>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZDRA4544YBXKE3XKBF76LTKBEFFANCNFSM427ILJUQ>
> .
>

rhatdan commented:
Let's see how bad the port to openshift builder is going to be.  If there is breakage, we can add back the parts that break as references to containers/libimage,

rhatdan commented:
/lgtm

vrothberg commented:
I looked at openshift/builder which does not build as I removed Push, Pull and changed the signature of util.FindImage.

I will spin up a follow-up PR tomorrow to add Push/Pull back. That's fairly trivial as I can move the logic in `cmd/buildah/....`. I'll also get the signature of FindImage back.
